state/store.go

```diff
@@ -651,14 +651,22 @@ func (store dbStore) LoadFinalizeBlockResponse(height int64) (*abci.FinalizeBloc
 
 	resp := new(abci.FinalizeBlockResponse)
 	err = resp.Unmarshal(buf)
+	// if an error is returned, then it needs to be handled here
 	if err != nil {
+		return nil, ErrFinalizeBlockResponseUnmarshalError
+	}
```

Comment:
So, the bug is that a `LegacyABCIResponses` struct _can_ be unmarshalled as `FinalizeBlockResponse` with no error (albeit with incomplete information), is that right?

Also, if seems that this logic wasn't really unit-tested (this is coming from famous #9468), otherwise we would have caught it. Could you please add a UT that ... ?
* fails without this PR
* passes with this PR
🙏

Reply:
OK, I will add a test

Reply:
I've implemented a test under `api/api_test.go`, this test tries to create a message at the v0.37 level using the same proto definitions and then convert it to a legacy ABCI response. I think this is a good way to test this use cases across proto versions.

I look into creating a test using the store methods SaveABCIResponseInfo but the way the code is structured in store a lot of things are internal and cannot be used in the test, we'd need to refactor things quite a bit. So focusing on the proto messages I think is a good way to test. I might look into testing something generating a message in v0.37 and using the byte buffer in a test, that would simulate some the storage too.